Issue/7227 tracking for cod

AnirudhBhat opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #7227
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds tracking for the Cash on Delivery project.

`card_present_onboarding_not_completed` with reason `cash_on_delivery_disabled`

**Button tap events**
**Skip:** add `card_present_onboarding_step_skipped` event, combined with a reason property matching above, and `remind_later` with a boolean (false for Iteration 1, may be true in Iteration 3). This can also be used for the Pending Requirements screen.

**Enable:** add `card_present_onboarding_cta_tapped`, combined with a reason property.


**Results**
**Enable succeeds:** new event `enable_cash_on_delivery_success`

**Enable fails:** new event `enable_cash_on_delivery_failed` with property error = "error message"

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### Onboarding not completed with reason
1. Ensure your store has COD disabled
2. Navigate to the IPP settings
3. Ensure you see COD disabled error screen
4. Check the logs to make sure you see `card_present_onboarding_not_completed` event with the reason as  `cash_on_delivery_disabled`


### Track COD enable Skip event
1. Ensure your store has COD disabled
2. Navigate to the IPP settings
3. Ensure you see COD disabled error screen
4. Tap on the Skip CTA
5. Check the logs to make sure you see `card_present_onboarding_step_skipped event`, combined with a reason property `cash_on_delivery_disabled`, and `remind_later` with a boolean (false for Iteration 1)

### Tracking Enable COD CTA tap
1. Ensure your store has COD disabled
2. Navigate to the IPP settings
3. Ensure you see COD disabled error screen
4. Tap on the Enable COD CTA
5. Check the logs to make sure you see `card_present_onboarding_cta_tapped event`, combined with a reason property `cash_on_delivery_disabled`


### Tracking enable COD success
1. Ensure your store has COD disabled
2. Navigate to the IPP settings
3. Ensure you see COD disabled error screen
4. Tap on the Enable COD CTA
5. Ensure the operation is success
6. Check the logs to make sure you see `enable_cash_on_delivery_success` event

### Tracking enable COD failure
1. Ensure your store has COD disabled
2. Navigate to the IPP settings
3. Ensure you see COD disabled error screen
4. Tap on the Enable COD CTA
5. Ensure the operation fails
6. Check the logs to make sure you see `enable_cash_on_delivery_failed` with property error = "error message"

### COD disabled screen when switching sites
1. Ensure your store has COD disabled
2. Navigate to the IPP settings
3. Ensure you see COD disabled error screen
4. Tap on skip
5. Ensure the COD disabled screen is not shown again for this site
6. Switch to some other site that has COD disabled
7. Navigate to IPP settings
8. Notice you see COD disabled screen.




- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

AnirudhBhat committed:
Add analytics event constants required for tracking in cash on delivery iteration 1

AnirudhBhat committed:
Add tracking in CardReaderTracker.kt

AnirudhBhat committed:
Integrate tracking in CardReaderOnboardingViewModel.kt

pachlava reviewed (approved):
I tested mentioned five scenarios with AS + Emulator, and saw the expected tracking messages in Logcat in all cases 👍

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/onboarding/CardReaderOnboardingViewModel.kt:
```diff
@@ -223,29 +223,76 @@ class CardReaderOnboardingViewModel @Inject constructor(
-                    onSkipCashOnDeliveryClicked = { (::onSkipCashOnDeliveryClicked)(state.countryCode) },
+                    onSkipCashOnDeliveryClicked = {
+                        (::onSkipCashOnDeliveryClicked)(
+                            state.countryCode,
```
❓ Have you considered sending `state` as parameter instead of splitting into fields and creating it again in the `onSkipCashOnDeliveryClicked` method? It feels weird that the VM is responsible for creating the onboardingState -> that's a responsibility of the OnboardingChecker, wdyt?

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/onboarding/CardReaderOnboardingViewModel.kt:
```diff
@@ -223,29 +223,76 @@ class CardReaderOnboardingViewModel @Inject constructor(
                     { (::onCashOnDeliveryEnabledSuccessfully)(state.countryCode) },
-                    onEnableCashOnDeliveryClicked = { (::onEnableCashOnDeliveryClicked)(state.countryCode) },
+                    onEnableCashOnDeliveryClicked = {
+                        (::onEnableCashOnDeliveryClicked)(
```
❓ Same as above.

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/onboarding/CardReaderOnboardingViewModel.kt:
```diff
@@ -223,29 +223,76 @@ class CardReaderOnboardingViewModel @Inject constructor(
+            )
+        )
         viewState.value = CashOnDeliveryDisabledState(
-            onSkipCashOnDeliveryClicked = { (::onSkipCashOnDeliveryClicked)(countryCode) },
```
❓ Have you considered creating a copy of the state instead of creating it from scratch?

`viewState.value = (viewState.value as CashOnDeliveryDisabledState).copy(shouldShowProgress = true)` ?

When the user clicks on the button we know for sure that the state is `CashOnDeliveryDisabledState` otherwise they wouldn't be able to click on it, right? Wdyt?

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/onboarding/CardReaderOnboardingViewModel.kt:
```diff
@@ -258,22 +305,50 @@ class CardReaderOnboardingViewModel @Inject constructor(
+                cardReaderTracker.trackCashOnDeliveryEnabledSuccess()
                 viewState.postValue(
                     CashOnDeliveryDisabledState(
-                        onSkipCashOnDeliveryClicked = { (::onSkipCashOnDeliveryClicked)(countryCode) },
```
❓  Same as above, have you considered creating a copy instead of creating the state from scratch?

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/onboarding/CardReaderOnboardingViewModel.kt:
```diff
@@ -258,22 +305,50 @@ class CardReaderOnboardingViewModel @Inject constructor(
+                )
                 viewState.postValue(
                     CashOnDeliveryDisabledState(
-                        onSkipCashOnDeliveryClicked = { (::onSkipCashOnDeliveryClicked)(countryCode) },
```
❓ Same as above.

malinajirka commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTrackerTest.kt:
```diff
@@ -115,6 +119,84 @@ class CardReaderTrackerTest : BaseUnitTest() {
+        }
+
+    @Test
+    fun `given enable cod is tapped, when success, then ENABLE_CASH_ON_DELIVERY_SUCCESS tracked`() =
```
🔍 It seems the given and when are swapped here. What about
`given enable cod request succeeds, when enable cod tapped, then ...`? Having said that, it feels weird that the test mentions UI even though the class is UI agnostic.

malinajirka commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTrackerTest.kt:
```diff
@@ -115,6 +119,84 @@ class CardReaderTrackerTest : BaseUnitTest() {
         }
 
+    @Test
+    fun `given cod disabled state, when skip is tapped, then CARD_PRESENT_ONBOARDING_STEP_SKIPPED tracked`() =
```
💡 It feels a bit weird that `CardReaderTracker` knows something about buttons - the class doesn't have any UI. We might want to consider making the test names more generic, wdyt?

malinajirka commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/onboarding/CardReaderOnboardingViewModel.kt:
```diff
@@ -223,29 +223,76 @@ class CardReaderOnboardingViewModel @Inject constructor(
                 viewState.value = CashOnDeliveryDisabledState(
-                    onSkipCashOnDeliveryClicked = { (::onSkipCashOnDeliveryClicked)(state.countryCode) },
+                    onSkipCashOnDeliveryClicked = {
+                        (::onSkipCashOnDeliveryClicked)(
```
❓ I'm not familiar with this syntax - what is the difference between the current code and
` { onEnableCashOnDeliveryClicked(...) }` ? If there isn't any, we might consider simplifying it, wdyt? (applies to several other places in this class)

AnirudhBhat committed:
Merge branch 'feature-cash-on-delivery' into issue/7227-tracking-for-cod

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/payments/cardreader/CardReaderTracker.kt

malinajirka reviewed (commented):
Thanks @AnirudhBhat!

I've reviewed the code and it seems to work as expected. I've left some comments but I'd consider fixing those in a follow up PR - later today/next week.


However, I noticed one minor bug -> `CARD_READER_DO_NOT_SHOW_CASH_ON_DELIVERY_DISABLED_ONBOARDING_STATE` shared preferences entry is not tied to a specific site but the whole app => we probably want to make it site specific (similarly as the other related preferences) so we show the COD error on each site not once per app installation (imo would be ideally fixed today or in the beta as otherwise we lose the information which sites the user has finished onboarding). Wdyt?

<img width="830" alt="Screenshot 2022-08-19 at 12 06 20" src="https://user-images.githubusercontent.com/2261188/185596161-1732d95a-6dfa-4972-b785-6f2cceb3f231.png">

AnirudhBhat committed:
Merge branch 'feature-cash-on-delivery' into issue/7227-tracking-for-cod

AnirudhBhat committed:
Fix a bug that caused to not show cod disabled screen on any of the sites during onboarding when skipped on one site.

AnirudhBhat commented:
Thanks for the review and thanks for catching the bug @malinajirka.

I've fixed the bug and now it should work across multiple sites. I've updated the PR description to include the test case as well. @pachlava  @rossanafmenezes  Can you please test the skip behavior when sites are switched?

AnirudhBhat committed:
Fix detekt errors

rossanafmenezes commented:
> I've fixed the bug and now it should work across multiple sites. I've updated the PR description to include the test case as well. @pachlava @rossanafmenezes Can you please test the skip behavior when sites are switched?

The screen shows now when you switch sites as expected if you skip the COD enable screen in one of the sites.
